obvspython/lenses.py

```diff
@@ -0,0 +1,191 @@
+"""
+lenses.py
+
+Implementation of some widely-known lenses
+"""
+
+from typing import List, Tuple
+import torch
+from plotly.graph_objects import Figure
+from obvspython.logging import logger
+from obvspython.vis import create_annotated_heatmap
+from obvspython.patchscope import SourceContext, TargetContext, Patchscope
+
+
+class BaseLogitLens:
+    """ Parent class for LogitLenses.
+        Patchscope and classic logit-lens are run differently,
+        but share the same visualization.
+     """
+
+    def __init__(self, model: str, prompt: str, device: str):
+        """ Constructor. Setup a Patchscope object with Source and Target context.
+            The target context is equal to the source context, apart from the layer.
+
+        Args:
+            model (str): Name of the model. Must be a valid name for huggingface transformers
+                package
+            prompt (str): The prompt to be analyzed
+            device (str): Device on which the model should be run: e.g. cpu, auto
+        """
+
+        self.model_name = model
+
+        # create SourceContext, leave position and layer as default for now
+        source_context = SourceContext(prompt=prompt, model_name=model, device=device)
+
+        # create TargetContext from SourceContext, as they are mostly equal
+        target_context = TargetContext.from_source(source_context)
+        # for the logit-lens, the target layer is always the last layer
+        target_context.layer = -1
+
+        # create Patchscope object
+        self.patchscope = Patchscope(source_context, target_context)
+        self.data = {}
+
+    def visualize(self, kind: str = 'top_logits_preds', file_name: str = '') -> Figure:
+        """ Visualize the logit lens results in one of the following ways:
+                top_logits_preds: Heatmap with the top predicted tokens and their logits
+        Args:
+              kind (str): The kind of visualization
+              file_name (str): If provided, save figure to a file with the given name
+                               (in the current path)
+        Returns (plotly.graph_objects.Figure):
+            The created figure
+        """
+
+        if not self.data:
+            logger.error('You need to call .run() before .visualize()!')
+            return Figure()
+
+        if kind == 'top_logits_preds':
+
+            logits = []
+            preds = []
+
+            # loop over all layers
+            for i, layer in enumerate(self.data['layers']):
+                layer_logits = []
+                layer_preds = []
+
+                # loop over every token in substring
+                for j in range(len(self.data['substring_tokens'])):
+
+                    # get the top prediction and logit
+                    top_logit, top_pred_idx = torch.max(self.data['logits'][(i, j)], dim=0)
```

Comment:
it doesnt matter much in this context, but two for loops for manipulating data in a torch tensor is kinda terrible. like, you could get the logits and preds from the tensor in one line, and then split them up.

basically the inner loop at least here should be replaced with tensor manipulation logic. looping over tensors and appending thier values to lists is just an extremely bad habit to get into when you dont have to

Reply:
I agree. The run method of the Patchscope logit-lens should definitely be improved.
I can do so in the next sprint.

Reply:
Aha, I mistook the code fragment. It's the visualization that's bad.
I will think about a better data structure here.